Fix re-triggering condition.

tigersoldier opened the pull request:
The current logic checks whether (point) changes before and after post
completion actions to determine if re-triggering completion is desired. However,
this logic is not robust enough. It incorrectly re-triggers completion when the
post-completion action adds import statements by Javacomp, or fixes the
duplicated `#` characters by cquery (see #21).

This change proposes comparing the suffix of the completion before and after
post-completion editing. I tested that it still works for the desired case
(`std::`) while not re-triggering completion for either the cquery include or
Javacomp auto-import case.

tigersoldier commented:
@Sarcasm @hicksy994 Would you like to review this PR?

Sarcasm commented:
I'm not 100% sure to understand the issue.
Shouldn't self completion be harmless if there is no trigger character to complete on?

After an include, I would not expect this to be a trigger point:

```cpp
#include "foo.h"|
```

tigersoldier commented:
I didn't read the cquery code. My guess is that cquery still thinks the completion prefix is `#include "foo.h"` and tries to complete it, resulting only one matching candidate with the same content.

To verify this theory, you can try invoking `M-x completion-at-point` or `M-x company-diag` at the end of the `#include` statement.

While the root-cause of the cquery bug is the mismatched prefix, it still has to be fixed for the Javacomp auto-import case. When I complete for `Gson`, Javacomp inserts `import com.google.gson.Gson` on the top of the file, but the completion itself isn't changed. However this triggers completion again because (point) has changed, showing candidates such as `GsonBuilder`.

Sarcasm commented:
Thank you for your detailed explanation.

Regarding cquery, I think the bug is in cquery.
But I find your issue for Javacomp legitimate.

TBH, I don't like the fix very much, it seems somewhat "fuzzy".

I'm thinking an alternative take, would be to use markers (https://www.gnu.org/software/emacs/manual/html_node/elisp/Markers.html#Markers).
Before the post-action, create a marker at point with `point-marker`.
Do the post-action.
If the point is on the marker, set the `self-insert` hack.
Remove marker.

What do you think?

tigersoldier commented:
I doubt if marker works at all. For the `textEdit` case, we backup the line before completion, restore it post-completion, then apply `textEdit` contents. I would imagine two possible results:

 1.  The marker is kept at `(point)` after applying `textEdit`. In this case, `std::` won't re-trigger completion.
 2.  The marker is not at `(point)`. In this case, if `textEdit` has the same content as `label` (cquery does that for variables), completion is re-triggered incorrectly.

 An alternative solution is to check if the suffix contains `triggerCharacters`. This is more reasonable and less hacky. What do you think?

Sarcasm commented:
> An alternative solution is to check if the suffix contains triggerCharacters. This is more reasonable and less hacky. What do you think?

Yes it looks more reasonable, that is actually what to achieve with my initial patch, to have triggers automatically enabled.

tigersoldier commented:
OK, changed the condition to use trigger characters.

I added a custom toggle for this feature and disabled it by default, because this solution doesn't fix the cquery include issue since `>' is one of the trigger characters. Although technically it's not our fault, I don't want to confuse cquery users until the bug is fixed on their side.

Sarcasm commented:
LGTM :+1:
